File: java20/src/main/java/com/syd/java20/problem/leetcode/LRUCache.java
package com.syd.java20.problem.leetcode;

import java.util.LinkedHashMap;

class LRUCache {
    private final int capacity;
    private final LinkedHashMap<Integer, Integer> map;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.map = new LinkedHashMap<>(capacity << 1, 0.75f, true);
    }

    public int get(int key) {return map.getOrDefault(key, -1);}

    public void put(int key, int value) {
        if (!map.containsKey(key) && map.size() >= capacity) map.remove(map.keySet().iterator().next());
        map.put(key, value);
    }
}

File: kt/src/main/java/com/behappy/kt/struct/LRUCache.kt
package com.behappy.kt.struct

import com.behappy.kt.invokeResults

class LRUCache(private val capacity: Int) {
    private val map = LinkedHashMap<Int, Int>(capacity.shl(1), 0.75f, true)
    fun get(key: Int) = map.getOrDefault(key, -1)
    fun put(key: Int, value: Int) {
        if (!map.containsKey(key) && map.size >= capacity) map.remove(map.entries.first().key)
        else map[key] = value
    }
}

fun main() {
    println(
        invokeResults(
            LRUCache::class,
            "[\"LRUCache\",\"put\",\"put\",\"get\",\"put\",\"get\",\"put\",\"get\",\"get\",\"get\"]",
            "[[2],[1,1],[2,2],[1],[3,3],[2],[4,4],[1],[3],[4]]"
        )
    )
}
